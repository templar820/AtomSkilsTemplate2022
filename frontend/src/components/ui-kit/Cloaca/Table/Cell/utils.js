import { useMemo } from 'react';

const useFixedDefiner = ({
    fixedHeader,
    fixedFooter,
    fixedFirstCol,
    fixedLastCol,
    firstCol,
    lastCol,
    tableSection,
    rowOffset
}) =>
    useMemo(() => {
        const result = {};
        if (fixedHeader && tableSection === 'thead') result.top = rowOffset !== undefined ? rowOffset : 'auto';
        if (fixedFooter && tableSection === 'tfoot') result.bottom = rowOffset !== undefined ? rowOffset : 'auto';
        if (fixedFirstCol && firstCol) result.left = 0;
        if (fixedLastCol && lastCol) result.right = 0;
        const length = Object.keys(result).length;
        if (length) {
            result.zIndex = length + (tableSection !== 'tbody' ? 2 : 1);
            return result;
        } else return false;
    }, [fixedHeader, fixedFooter, fixedFirstCol, fixedLastCol, firstCol, lastCol, rowOffset]);

export { useFixedDefiner };
