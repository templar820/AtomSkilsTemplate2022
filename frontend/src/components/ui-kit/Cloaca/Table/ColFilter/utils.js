import React, { useCallback, useMemo } from 'react';

const useMapHeadings = (rows) =>
    useMemo(() => {
        const array = React.Children.toArray(rows);
        if (!array.length) return [];
        const { children } = array[0].props || { children: [] };
        return Array.from(
            children.reduce(
                (headings, { props: { children, showInColFilter } = {} } = {}, index) =>
                    showInColFilter ? headings.set(index, children) : headings,
                new Map()
            )
        );
    }, [rows]);

const useHandlerCreator = (setFilteredColumns) =>
    useCallback(
        (id) =>
            setFilteredColumns((prevFilteredColumns) => {
                const index = prevFilteredColumns.findIndex((prevId) => prevId === id);
                const newFilteredColumns = prevFilteredColumns.slice();
                if (index === -1) {
                    newFilteredColumns.push(id);
                } else {
                    newFilteredColumns.splice(index, 1);
                }
                return newFilteredColumns;
            }),
        [setFilteredColumns]
    );

export { useMapHeadings, useHandlerCreator };
