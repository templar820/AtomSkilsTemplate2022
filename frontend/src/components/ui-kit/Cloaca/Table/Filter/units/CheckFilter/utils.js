const createOnCheckHandler = ({
    updateFilter,
    filterOnOptionPick,
    revertOnCancel,
    uniques,
    setValue,
    onChange,
    customRender
}) => (id) => {
    if (customRender && typeof onChange === 'function') return onChange(id);

    setValue((prevValue = uniques) => {
        const value = prevValue.includes(id) ? prevValue.filter((item) => item !== id) : [...prevValue, id];
        filterOnOptionPick && updateFilter(value, revertOnCancel);
        return value;
    });
};

const createOnOkHandler = ({ updateFilter, value, closeOnOk, onOk }) => ({ setIsOpenMain }) => {
    if (typeof onOk === 'function') onOk({ value });
    updateFilter(value, false);
    closeOnOk && setIsOpenMain(false);
};

const createOnCancelHandler = ({ updateFilter, revertOnCancel, closeOnCancel, stableFilter, onCancel }) => ({
    setIsOpenMain
}) => {
    if (typeof onCancel === 'function') onCancel();
    if (revertOnCancel) updateFilter(stableFilter.current);
    if (closeOnCancel) setIsOpenMain(false);
};

export { createOnCheckHandler, createOnCancelHandler, createOnOkHandler };
