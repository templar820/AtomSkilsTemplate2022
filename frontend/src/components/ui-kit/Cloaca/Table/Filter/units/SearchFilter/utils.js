const createOnChangeHandler = ({ updateFilter, filterOnOptionPick, revertOnCancel, setValue, value }) => (values) => {
    const newValue = values.length ? values[0].label : undefined;
    if (newValue !== value) {
        setValue(newValue);
        filterOnOptionPick && updateFilter(newValue, revertOnCancel);
    }
};

const createOnOkHandler = ({ updateFilter, value, closeOnOk, onOk }) => (e, dropdownControls = {}, meta) => {
    if (typeof onOk === 'function') onOk({ value, meta });
    updateFilter(value, false);
    closeOnOk && dropdownControls.setIsOpenMain && dropdownControls.setIsOpenMain(false);
};

const createClearAllHandler = ({ updateFilter, revertOnCancel, closeOnCancel, stableFilter, onCancel }) => (
    e,
    dropdownControls = {}
) => {
    if (typeof onCancel === 'function') onCancel();
    if (revertOnCancel) {
        updateFilter(stableFilter.current);
        closeOnCancel && dropdownControls.setIsOpenMain && dropdownControls.setIsOpenMain(false);
    }
};

const createOnSearchChangeHandler = ({ updateFilter, filterOnType, revertOnCancel, setValue }) => ({}, { search }) => {
    setValue(search);
    filterOnType && updateFilter(search, revertOnCancel);
};

const getDropdownSettings = (showDropdown) =>
    !showDropdown
        ? {
              dropdownProps: {
                  css: { display: 'none' }
              },
              dropdownRenderer: () => null
          }
        : {};

export {
    createClearAllHandler,
    createOnChangeHandler,
    createOnOkHandler,
    createOnSearchChangeHandler,
    getDropdownSettings
};
