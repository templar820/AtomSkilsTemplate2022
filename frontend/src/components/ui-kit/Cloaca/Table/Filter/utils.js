import { useCallback } from 'react';

import filtersCreators from 'lib-ui/Table/filters';

const useFilterUpdaterCreator = ({
    setFilters,
    stableFilter,
    filterType,
    colIndex,
    prevValue,
    onChange,
    preventFiltersUpdating
}) =>
    useCallback(
        (value, temp = false) => {
            const handleOnChange = () => {
                if (typeof onChange === 'function') onChange({ value });
            };

            if (preventFiltersUpdating) {
                handleOnChange();
                return;
            }

            if (typeof value === 'function') {
                filterType = value.filterType;
                value = value.compare;
            }

            handleOnChange();

            const newFilter = (value || value === 0) && filtersCreators[filterType](colIndex, value, temp);
            if (!temp) stableFilter.current = newFilter;
            setFilters((filters) => {
                const index = filters.findIndex((filter) => filter.byColumn === colIndex);
                if (newFilter !== false) {
                    return index >= 0
                        ? filters.map((filter, filterIndex) => (filterIndex !== index ? filter : newFilter))
                        : [...filters, newFilter];
                } else {
                    return filters.filter((filter, filterIndex) => filterIndex !== index);
                }
            });
        },
        [setFilters, stableFilter, filterType, colIndex, prevValue, onChange]
    );

export { useFilterUpdaterCreator };
