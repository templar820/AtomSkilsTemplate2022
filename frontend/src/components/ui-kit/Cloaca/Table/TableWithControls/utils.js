import React, { useCallback, useMemo, useState, useRef } from 'react';
import { defaultFirstCollCells, defaultLastCollCells } from './units';

const insertCell = ({ children, cell = undefined, position = 'last' }) => {
    const cells = React.Children.toArray(children);
    if (cell) cells.splice(position === 'last' ? cells.length : position, 0, cell);
    return cells;
};

const usePropsInsert = ({ children, addFirstCol, addLastCol, firstColCells, lastColCells }) =>
    useMemo(() => {
        if (!addFirstCol && !addLastCol) return children;
        const totalFirstColCells = { ...defaultFirstCollCells, ...firstColCells };
        const totalLastColCells = { ...defaultLastCollCells, ...lastColCells };
        return React.Children.map(children, (child) => {
            const { tableSection } = child.type;
            const { children, ...restChildProps } = child.props;
            const rows = React.Children.map(children, (row, index) => {
                const { props: { children } = {}, key } = row || {};
                const cellsWithLastCell = !addLastCol
                    ? children
                    : insertCell({ children, cell: totalLastColCells[tableSection], position: 'last' });
                const cellsWithFirstCell = !addFirstCol
                    ? cellsWithLastCell
                    : insertCell({
                          children: cellsWithLastCell,
                          cell: totalFirstColCells[tableSection],
                          position: 0
                      });
                const originalRowKey = key || String(index);
                return { ...row, props: { ...row.props, children: cellsWithFirstCell, originalRowKey } };
            });
            return React.cloneElement(
                child,
                {
                    ...restChildProps
                },
                rows
            );
        });
    }, [children, addFirstCol, addLastCol, firstColCells, lastColCells]);

const getColumnsCount = (children) => {
    let section = React.Children.toArray(children).find(
        (child) => child.props && child.props.children && React.Children.count(child.props.children)
    );
    if (!section) {
        throw new Error('No <Row/> found');
    }
    const row = React.Children.toArray(section.props.children)[0] || {};
    return row.props && row.props.children
        ? row.props.children.reduce((acc, cell) => {
              if (cell && cell.props) acc += cell.props.colSpan ? cell.props.colSpan : 1;
              return acc;
          }, 0)
        : 0;
};

const calcStartEntry = (page, pagesCount, rowsCount, extendLastPage, entriesOnDisplay) =>
    page !== pagesCount || !extendLastPage ? (page - 1) * entriesOnDisplay : rowsCount - entriesOnDisplay;

const getRowSelectableStatus = (isSelectable, tableSection) => {
    return typeof isSelectable === 'boolean' ? isSelectable : tableSection === 'tbody';
};

const getObjectForAllCheckboxes = (isCheckingControlled, checkedRows, mergeCheckedRows) => {
    switch (true) {
        case checkedRows === undefined:
            return {};
        case typeof checkedRows !== 'object' ||
            Object.keys(checkedRows).some((key) => key !== 'tbody' && key !== 'thead' && key !== 'foot'):
            // dealing with wrong prop structure
            console.error(
                'prop checkedRows is defined with errors. checkedRows must be an object with table sections keys.'
            );
            return {};
        case !mergeCheckedRows:
            return checkedRows;
        default:
            // prevent mutating of original objects
            return Object.entries(checkedRows).reduce((result, [section, keys]) => {
                result[section] = { ...keys };
                return result;
            }, {});
    }
};

const useCheckController = (
    children,
    isCheckingControlled,
    checkedRows,
    defaultChecked,
    mergeCheckedRows,
    callback
) => {
    const currentCheckboxesRef = useRef(null);
    currentCheckboxesRef.current = useMemo(() => {
        // ref used to prevent expensive calculations in non-controlled mode
        if (!isCheckingControlled && currentCheckboxesRef.current !== null) return currentCheckboxesRef.current;

        const obj = getObjectForAllCheckboxes(isCheckingControlled, checkedRows, mergeCheckedRows);
        if (obj === checkedRows) return checkedRows;

        // iterations over sections
        React.Children.forEach(children, (child) => {
            const { tableSection } = child.type;

            // iterations over rows
            React.Children.forEach(child.props.children, ({ props: { originalRowKey: key, isSelectable } = {} }) => {
                const finalIsSelectable = getRowSelectableStatus(isSelectable, tableSection);
                if (finalIsSelectable) {
                    if (!obj[tableSection]) obj[tableSection] = {};
                    if (obj[tableSection][key] === undefined) obj[tableSection][key] = defaultChecked;
                }
            });
        });

        return obj;
    }, [children, isCheckingControlled, checkedRows, defaultChecked, mergeCheckedRows]);

    const [checkState, setCheckState] = useState(currentCheckboxesRef.current);
    const allCheck = isCheckingControlled ? currentCheckboxesRef.current : checkState;

    const setAllCheck = useCallback(
        (newState, change) => {
            if (!isCheckingControlled) setCheckState(newState);
            callback && callback(newState, change);
        },
        [callback, isCheckingControlled]
    );

    return { allCheck, setAllCheck };
};

const getTotalChecked = (allCheck) => {
    let result = undefined;
    const sections = Object.keys(allCheck);
    // iterate overs sections
    for (let i = 0; result !== 'indeterminate' && i < sections.length; i++) {
        const rows = Object.values(allCheck[sections[i]]);
        // iterate over rows in section
        for (let j = 0; result !== 'indeterminate' && j < rows.length; j++) {
            if (result !== undefined) {
                result = result === rows[j] ? rows[j] : 'indeterminate';
            } else {
                result = rows[j];
            }
        }
    }
    return result !== undefined ? result : false;
};

const handleCheck = ({ tableSection, setAllCheck, allCheck, originalRowKey, sumCell = false, event }) => {
    const obj = { ...allCheck };
    const value = event.target.checked;

    if (sumCell) {
        // iterate over sections
        for (let section in obj) {
            if (obj.hasOwnProperty(section)) {
                // iterate over rows
                for (let key in obj[section]) {
                    if (obj[section].hasOwnProperty(key)) obj[section][key] = value;
                }
            }
        }
    } else {
        obj[tableSection][originalRowKey] = value;
    }

    setAllCheck(obj, { key: originalRowKey, tableSection, value, sumCell });
};

const fillArrayOfDefaultValues = (array, defaultValue, count) => {
    return Array.apply(null, Array(count)).reduce(
        (acc, el, i) => [...acc, array[i] === undefined ? defaultValue : array[i]],
        []
    );
};

const getArrayWithAddedColumns = (array, count, isFirst, isLast, firstValue, lastValue, defaultValue) => {
    let tempArr = isFirst ? [firstValue, ...array] : array;
    if (isLast) {
        tempArr = fillArrayOfDefaultValues(tempArr, defaultValue, count);
        tempArr[tempArr.length - 1] = lastValue;
    }
    return tempArr;
};

const prepareVirtualization = (virtualization, onPageChange, entriesOnDisplay) => {
    if (!virtualization) return false;
    return {
        ...(virtualization || {}),
        onScroll: (arg) => {
            if (virtualization.onScroll) virtualization.onScroll(arg);
            onPageChange(Math.floor(arg.calcPureStartEntry() / entriesOnDisplay) + 1);
        }
    };
};

export {
    getColumnsCount,
    getArrayWithAddedColumns,
    calcStartEntry,
    insertCell,
    usePropsInsert,
    handleCheck,
    getTotalChecked,
    useCheckController,
    prepareVirtualization
};
