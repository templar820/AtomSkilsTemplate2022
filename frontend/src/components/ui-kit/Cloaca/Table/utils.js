import React from 'react';

const defaultControls = {};
const ControlsContext = React.createContext(defaultControls);

// 1 - startEntry. Offset as items count
// 2 - Limit. Number of items to show
// 3 - offset. Space that compensate non-showed items from the start of the list (set by startEntry).
const OffsetContext = React.createContext([0, 10, 0]);

const getCellValue = (row, colIndex) => {
    const { props: { filtersValue, children } = {} } = React.Children.toArray(row.props.children)[colIndex] || {};
    if (filtersValue !== undefined) return filtersValue;
    return typeof children === 'string' || typeof children === 'number' ? children : 0;
};

export { getCellValue, ControlsContext, OffsetContext };
