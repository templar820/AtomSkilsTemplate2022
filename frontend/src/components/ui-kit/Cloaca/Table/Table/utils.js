import React, { useState, useEffect, useMemo, useCallback, useLayoutEffect, useRef } from 'react';
import { isEqual, useRefCallback, useResizeObserver, useStateController, useObjectMemo } from 'lib-root/utils';

const filterList = (list, filteredMap) => list.filter((_, index) => !filteredMap[index]);
const fillList = (count, widthsList, filler) =>
    [...new Array(count)].map((_, index) => (widthsList[index] !== undefined ? widthsList[index] : filler));
const useFillListMemo = (count, filteredMap, list, filler) =>
    useMemo(() => {
        const filledList = fillList(count, list, filler);
        const filteredList = filterList(filledList, filteredMap);
        return [filteredList, filledList];
    }, [count, list, filler]);

const useResizerStates = (onResizeStart, onResizeEnd) => {
    // make it this way to prevent too frequent invalidation of useCallback
    const onResizeStartRef = useRef(onResizeStart);
    onResizeStartRef.current = onResizeStart;
    const onResizeEndRef = useRef(onResizeEnd);
    onResizeEndRef.current = onResizeEnd;

    const [resizingCols, setResizingCols] = useState(null);
    const [isResizing, setIsResizing] = useState(false);
    const seIsResizingWithCallback = useCallback(
        (newValue) => {
            setIsResizing(() => {
                const callback = (newValue ? onResizeStartRef : onResizeEndRef).current;
                if (callback) callback();
                return newValue;
            });
        },
        [setIsResizing]
    );
    return useObjectMemo({ resizingCols, setResizingCols, isResizing, setIsResizing: seIsResizingWithCallback });
};

const splitChildren = (children) =>
    Object.assign(
        ...React.Children.map(
            children,
            ({ type: { tableSection }, props: { children } }) => {
                return { [tableSection]: children };
            },
            null
        )
    );
const filterRows = (grandchildren, filters) =>
    !filters.length
        ? grandchildren
        : {
              ...grandchildren,
              tbody: filters.reduce((rows, filter, index) => filter(rows, index), grandchildren.tbody)
          };
const filterCols = (grandchildren, filteredColumnsMap, totalColumnsCount) => {
    let columnsCount = 0;

    const filteredGrandchildren = Object.assign(
        ...Object.keys(grandchildren).map((section) => {
            const rows = React.Children.map(
                grandchildren[section],
                (row) => {
                    const { children } = row.props || {};
                    const childrenArray = React.Children.toArray(children);
                    const filteredChildren = !filteredColumnsMap.length
                        ? childrenArray
                        : childrenArray.filter(({ props: { index } = {} }) => !filteredColumnsMap[index]);
                    if (filteredChildren.length > columnsCount) columnsCount = filteredChildren.length;
                    return { ...row, props: { ...row.props, children: filteredChildren } };
                },
                null
            );
            return { [section]: rows };
        })
    );
    if (!filteredColumnsMap.length) columnsCount = totalColumnsCount;

    return [filteredGrandchildren, columnsCount];
};
const higherSort = (a, b, sorters, index) => {
    try {
        return sorters[index](a, b) || (sorters[index + 1] && higherSort(a, b, sorters, index + 1));
    } catch (e) {
        console.error(e);
        return 0;
    }
};
const reorderSorters = (sortersOrder, sorters) => {
    switch (true) {
        case sortersOrder === 'ltr':
            return [...sorters].sort((a, b) => a['byColumn'] - b['byColumn']);
        case sortersOrder === 'rtl':
            return [...sorters].sort((a, b) => b['byColumn'] - a['byColumn']);
        case Array.isArray(sortersOrder):
            const orderedPart = sortersOrder.map((position) => sorters[position]);
            const otherSorters = sorters.filter((sorter, index) => !sortersOrder.includes(index));
            return [...orderedPart, ...otherSorters].filter((sorter) => !!sorter);
        case typeof sortersOrder === 'function':
            return sortersOrder(sorters);
        default:
            return sorters;
    }
};
const sortRows = (grandchildren, sorters, sortersOrder) => {
    if (!sorters.length) return grandchildren;
    const reorderedSorters = reorderSorters(sortersOrder, sorters);
    return !reorderedSorters.length
        ? grandchildren
        : {
              ...grandchildren,
              tbody: [...grandchildren.tbody].sort((a, b) => higherSort(a, b, reorderedSorters, 0))
          };
};

/**
 * Returns current state and then decreases 'rowSpan' field of each child of 'spans'
 * @param spans - {{[index]: {colSpan, rowSpan}, ...}}
 * @returns {{[index]: {colSpan, rowSpan}, ...}}
 */
const decreaseRowSpans = (spans = {}) => {
    const prevState = { ...spans };

    for (let index in spans) {
        spans[index].rowSpan--;
        if (spans[index].rowSpan <= 1) delete spans[index];
    }

    return prevState;
};

const indexColumns = (grandchildren) => {
    let columnsCount = 0;
    const spans = {};

    // looping table sections
    const indexedGrandchildren = Object.assign(
        ...Object.keys(grandchildren).map((grandchild) => ({
            // mapping rows
            [grandchild]: React.Children.map(
                grandchildren[grandchild],
                (row) => {
                    const { children = [] } = row.props || {};
                    const cells = React.Children.toArray(children);
                    const _spans = decreaseRowSpans(spans);

                    // mapping cells:
                    const { indexedCells, colIndex: rowColumnsCount } = cells.reduce(
                        ({ indexedCells, colIndex }, cell) => {
                            const { rowSpan = 1, colSpan = 1 } = cell.props || {};
                            let modifier = 0;

                            if (_spans[colIndex]) {
                                modifier = _spans[colIndex].colSpan || 0;
                            }

                            const resultIndex = modifier + colIndex;

                            if (rowSpan > 1) spans[resultIndex] = { rowSpan, colSpan };

                            const resultCell = React.cloneElement(cell, { index: resultIndex });

                            return { indexedCells: [...indexedCells, resultCell], colIndex: resultIndex + colSpan };
                        },
                        { indexedCells: [], colIndex: 0 }
                    );

                    // update columnsCount
                    if (rowColumnsCount > columnsCount) columnsCount = rowColumnsCount;

                    return React.cloneElement(row, null, indexedCells);
                },
                null
            )
        }))
    );

    return { totalColumnsCount: columnsCount, indexedGrandchildren };
};

const useGrandchildrenPreparing = (children, filters, sorters, sortersOrder, filteredColumnsMap) =>
    useMemo(() => {
        {
            const grandchildren = splitChildren(children);
            const { totalColumnsCount, indexedGrandchildren } = indexColumns(grandchildren);
            const filteredGrandchildren = filterRows(indexedGrandchildren, filters);
            const [colFilteredGrandchildren, columnsCount] = filterCols(
                filteredGrandchildren,
                filteredColumnsMap,
                totalColumnsCount
            );
            const sortedGrandchildren = sortRows(colFilteredGrandchildren, sorters, sortersOrder);
            // const { columnsCount, indexedGrandchildren } = indexColumns(sortedGrandchildren);
            return {
                grandchildren: sortedGrandchildren,
                allGrandchildren: grandchildren,
                totalColumnsCount,
                columnsCount
            };
        }
    }, [children, filters, sorters, sortersOrder, filteredColumnsMap]);

const getScrolls = (
    { scrollTop, scrollLeft, scrollHeight, scrollWidth, clientHeight, clientWidth },
    [fixedHeader, fixedFooter, fixedFirstCol, fixedLastCol]
) => ({
    top: !!(scrollTop && fixedHeader),
    right: !!(fixedLastCol && scrollWidth - clientWidth - scrollLeft),
    bottom: !!(scrollHeight - clientHeight - scrollTop && fixedFooter),
    left: !!(fixedFirstCol && scrollLeft)
});

const useScrollDeterminer = (...fixedPositions) => {
    const { ref: wrapperRef, initRef: setWrapperRef } = useRefCallback();
    const [isScrolled, setIsScrolled] = useState({
        top: false,
        right: false,
        bottom: false,
        left: false
    });
    const handleScroll = useCallback(
        () =>
            setIsScrolled((state) => {
                const newState = getScrolls(wrapperRef.current, fixedPositions);
                return isEqual(state, newState) ? state : newState;
            }),
        [wrapperRef.current, ...fixedPositions]
    );

    const shouldListen = fixedPositions.some((position) => position);

    // check scrolls when table resized
    useResizeObserver({ ref: shouldListen ? wrapperRef : { current: null }, callback: handleScroll });
    // check scrolls when table scrolled
    useEffect(() => {
        if (wrapperRef.current && shouldListen) {
            handleScroll();

            wrapperRef.current.addEventListener('scroll', handleScroll);
            return () => wrapperRef.current.removeEventListener('scroll', handleScroll);
        }
    }, [wrapperRef.current, handleScroll]);

    return { setWrapperRef, wrapperRef, isScrolled };
};

const useRowsOffsetsGetter = ({ current: table }) => {
    const [heights, setHeights] = useState([]);
    useLayoutEffect(() => {
        if (table) {
            const rows = Array.prototype.slice.call(table.querySelectorAll('tr'));
            const newHeights = rows.map((row) => row.offsetHeight);
            setHeights(isEqual(newHeights, heights) ? heights : newHeights);
        }
    });

    return useMemo(() => {
        const tops = heights.reduce((result, height, index) => [...result, result[index] + height], [0]);
        const bottoms = tops.map((item) => tops[tops.length - 1] - item);
        return {
            tops,
            bottoms
        };
    }, [heights]);
};

const useFilteredColumnsController = (propFilteredColumns, colFilters, callback, isControlled) => {
    const value = useMemo(() => {
        if (!Array.isArray(colFilters)) {
            return propFilteredColumns || [];
        } else {
            return colFilters.reduce((result, { compare }) => {
                if (result.indexOf(compare) === -1) result.push(compare);
                return result;
            }, []);
        }
    }, [propFilteredColumns, colFilters]);

    const { value: filteredColumns, handler: setFilteredColumns } = useStateController({
        value,
        callback,
        isControlled
    });
    const filteredColumnsMap = useMemo(
        () =>
            filteredColumns.reduce((map, columnIndex) => {
                map[columnIndex] = true;
                return map;
            }, []),
        [filteredColumns]
    );
    return [filteredColumnsMap, filteredColumns, setFilteredColumns];
};

export {
    useResizerStates,
    useScrollDeterminer,
    useGrandchildrenPreparing,
    filterList,
    useFillListMemo,
    indexColumns,
    useRowsOffsetsGetter,
    useFilteredColumnsController
};
