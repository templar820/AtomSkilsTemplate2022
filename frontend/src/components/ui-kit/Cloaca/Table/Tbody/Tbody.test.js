import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { matchers } from 'jest-emotion';
import serializer from 'jest-emotion';
import { Table, Tbody, Row } from '../';
import { testStyles } from 'lib-root/utils';

Enzyme.configure({ adapter: new Adapter() });
expect.extend(matchers);
expect.addSnapshotSerializer(serializer);

const props = {
    css: { color: '#fff' }
};

describe('<Tbody />', () => {
    let table, tbody;

    beforeEach(() => {
        table = mount(
            <Table>
                <Tbody {...props}>
                    <Row></Row>
                </Tbody>
            </Table>
        );

        tbody = table.find(Tbody);
    });

    afterEach(() => {
        table.unmount();
    });

    test('should match snapshot', () => {
        expect(toJson(tbody)).toMatchSnapshot();
    });

    test('should render correctly', () => {
        testStyles(tbody, props.css);
    });
});
