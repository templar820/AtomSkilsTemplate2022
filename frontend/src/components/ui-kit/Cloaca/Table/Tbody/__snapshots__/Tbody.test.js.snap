// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tbody /> should match snapshot 1`] = `
.emotion-2 {
  color: #fff;
}

.emotion-2:after {
  content: '';
  display: table-row;
  height: 100%;
}

.emotion-0 {
  background-color: #ffffff;
}

.emotion-0:hover,
.emotion-0.selected {
  background-color: #f6f6f6;
}

<ForwardRef
  colsStyles={Array []}
  columnsCount={0}
  coords={null}
  css={
    Object {
      "color": "#fff",
    }
  }
  fixedFirstCol={false}
  fixedFooter={false}
  fixedHeader={true}
  fixedLastCol={false}
  hoveringRows={true}
  isScrolled={
    Object {
      "bottom": false,
      "left": false,
      "right": false,
      "top": false,
    }
  }
  key=".0"
  noDataProps={
    Object {
      "text": "",
    }
  }
  resizableCols={Array []}
  rowsOffsets={
    Object {
      "bottoms": Array [
        0,
        0,
        0,
      ],
      "tops": Array [
        0,
        0,
        0,
      ],
    }
  }
  setCoords={[Function]}
  showHorizontalScrollShadows={true}
  showNoData={true}
  tableSection="tbody"
>
  <StyledTbody
    _css={
      Object {
        "color": "#fff",
      }
    }
    className="CLR-table__body"
    colors={
      Object {
        "GrayScale_0": "#ffffff",
        "GrayScale_100": "#EDEDED",
        "GrayScale_150": "#E5E5E5",
        "GrayScale_200": "#D8D8D8",
        "GrayScale_50": "#fafafa",
        "GrayScale_500": "#999999",
        "GrayScale_55": "#f6f6f6",
        "GrayScale_600": "#7C7C7C",
        "GrayScale_700": "#333333",
        "GrayScale_900": "#000000",
        "error": "#F62434",
        "errorAccent": "#EB5A40",
        "info": "#34AAF2",
        "infoAccent": "#5BD9E5",
        "primary": "#00B956",
        "primaryAccent": "#10E272",
        "secondary": "#731982",
        "secondaryAccent": "#A500BF",
        "tertiary": "#444189",
        "tertiaryAccent": "#404D46",
        "warning": "#FFA717",
        "warningAccent": "#F36F20",
      }
    }
  >
    <tbody
      className="CLR-table__body emotion-2 emotion-3"
    >
      <ForwardRef
        colsStyles={Array []}
        columnsCount={0}
        coords={null}
        fixedFirstCol={false}
        fixedFooter={false}
        fixedHeader={true}
        fixedLastCol={false}
        hoveringRows={true}
        isScrolled={
          Object {
            "bottom": false,
            "left": false,
            "right": false,
            "top": false,
          }
        }
        key=".$.0"
        resizableCols={Array []}
        rowsOffsets={
          Object {
            "bottoms": Array [
              0,
              0,
              0,
            ],
            "tops": Array [
              0,
              0,
              0,
            ],
          }
        }
        setCoords={[Function]}
        showHorizontalScrollShadows={true}
        tableSection="tbody"
      >
        <StyledRow
          className="CLR-table__row"
          colors={
            Object {
              "GrayScale_0": "#ffffff",
              "GrayScale_100": "#EDEDED",
              "GrayScale_150": "#E5E5E5",
              "GrayScale_200": "#D8D8D8",
              "GrayScale_50": "#fafafa",
              "GrayScale_500": "#999999",
              "GrayScale_55": "#f6f6f6",
              "GrayScale_600": "#7C7C7C",
              "GrayScale_700": "#333333",
              "GrayScale_900": "#000000",
              "error": "#F62434",
              "errorAccent": "#EB5A40",
              "info": "#34AAF2",
              "infoAccent": "#5BD9E5",
              "primary": "#00B956",
              "primaryAccent": "#10E272",
              "secondary": "#731982",
              "secondaryAccent": "#A500BF",
              "tertiary": "#444189",
              "tertiaryAccent": "#404D46",
              "warning": "#FFA717",
              "warningAccent": "#F36F20",
            }
          }
          hoveringRows={true}
          onContextMenu={[Function]}
          tableSection="tbody"
        >
          <tr
            className="CLR-table__row emotion-0 emotion-1"
            onContextMenu={[Function]}
          />
        </StyledRow>
      </ForwardRef>
    </tbody>
  </StyledTbody>
</ForwardRef>
`;
